C error on assign aliased fixed array to option variable

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

**V doctor:**

```
V full version: V 0.4.8 5ec9bb5.1fae506
OS: linux, Linux version 6.6.8-arch1-1 (linux@archlinux) (gcc (GCC) 13.2.1 20230801, GNU ld (GNU Binutils) 2.41.0) #1 SMP PREEMPT_DYNAMIC Thu, 21 Dec 2023 19:01:01 +0000
Processor: 16 cpus, 64bit, little endian, 12th Gen Intel(R) Core(TM) i5-1240P

getwd: /home/ge
vexe: /home/ge/.vlang/v
vexe mtime: 2024-12-16 20:49:32

vroot: OK, value: /home/ge/.vlang
VMODULES: OK, value: /home/ge/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.47.1
Git vroot status: weekly.2024.50-49-g1fae5069
.git/config present: true

CC version: cc (GCC) 14.2.1 20240910
emcc version: N/A
thirdparty/tcc status: thirdparty-linux-amd64 0134e9b9
```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg aliased_fixed_array_option_assign.vv`

```v
type Addr = [4]u8

mut addr := ?Addr(none)

addr = Addr([u8(1), 2, 3, 4]!)

println(addr)
```

**What did you expect to see?**

`Option(Addr([1, 2, 3, 4]))` printed. This works for non-fixed arrays:

```v
type Addr = []u8

mut addr := ?Addr(none)

addr = Addr([u8(1), 2, 3, 4])

println(addr)
// Option(Addr([1, 2, 3, 4]))
```

**What did you see instead?**

```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/aliased_fixed_array_option_assign.01JF8KQTJXZPBWEQ5FYB0PBK9N.tmp.c:5845: error: cannot convert 'struct _option_main__Addr' to 'void *'
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzYwOTMxNGYxNDRmNTg1YWU0ZGExNDIiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.Oy5t2YWRZDFt0QyplQr5js2wROSxz5-H43YP4-5BjY4">Huly&reg;: <b>V_0.6-21622</b></a></sub>

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -492,11 +492,21 @@ fn (mut g Gen) assign_stmt(node_ ast.AssignStmt) {
 				g.write(' = ')
 				g.expr_with_opt(val, val_type, var_type)
 			} else if unaliased_right_sym.kind == .array_fixed && val is ast.CastExpr {
-				g.write('memcpy(')
-				g.expr(left)
-				g.write(', ')
-				g.expr(val)
-				g.writeln(', sizeof(${g.styp(var_type)}));')
+				if var_type.has_flag(.option) {
+					g.expr(left)
+					g.writeln('.state = 0;')
+					g.write('memcpy(')
+					g.expr(left)
+					g.write('.data, ')
+					g.expr(val)
+					g.writeln(', sizeof(${g.styp(var_type.clear_flag(.option))}));')
+				} else {
+					g.write('memcpy(')
+					g.expr(left)
+					g.write(', ')
+					g.expr(val)
+					g.writeln(', sizeof(${g.styp(var_type)}));')
+				}
 			} else {
 				mut v_var := ''
 				arr_typ := styp.trim('*')

--- a/vlib/v/tests/assign/assign_option_of_fixed_array_test.v
+++ b/vlib/v/tests/assign/assign_option_of_fixed_array_test.v
@@ -1,0 +1,8 @@
+type Addr = [4]u8
+
+fn test_assign_option_of_fixed_array() {
+	mut addr := ?Addr(none)
+	addr = Addr([u8(1), 2, 3, 4]!)
+	println(addr)
+	assert '${addr}' == 'Option(Addr([1, 2, 3, 4]))'
+}

